search/agentspace/agentspace-sfdc-workforce-idp-acl/agentspace-sfdc-workfoce-acl.py

```diff
@@ -0,0 +1,247 @@
+        query_params = parse_qs(urlparse(self.path).query)
+
+        if "code" in query_params and "state" in query_params:
+            with server_lock:
+                auth_code = query_params["code"][0]
+                received_state = query_params["state"][0]
+            self.send_response(200)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+            self.wfile.write(b"Authentication successful! Close this window.")
+        else:
+            self.send_response(400)
+            self.send_header("Content-type", "text/html")
+            self.end_headers()
+            self.wfile.write(b"Error: Missing code or state.")
+
+
+def run_server():
+    global server_ready
+    httpd = HTTPServer(("localhost", 8000), CallbackHandler)
+    with server_lock:
+        server_ready = True
+    httpd.handle_request()
+    httpd.server_close()
+
+
+# --- Helper Functions for PKCE ---
+def generate_code_verifier():
+    return secrets.token_urlsafe(64)
+
+
+def generate_code_challenge(code_verifier):
+    hashed = hashlib.sha256(code_verifier.encode("utf-8")).digest()
+    return base64.urlsafe_b64encode(hashed).decode("utf-8").rstrip("=")
+
+
+# --- Main Functions ---
+def get_salesforce_token():
+    global auth_code, received_state, server_ready, code_verifier
+    state = secrets.token_urlsafe(16)
+    nonce = secrets.token_urlsafe(16)
+    server_ready = False
+    auth_code = None
+    received_state = None
+
+    code_verifier = generate_code_verifier()
+    code_challenge = generate_code_challenge(code_verifier)
+
+    server_thread = Thread(target=run_server)
+    server_thread.start()
+
+    while not server_ready:
+        time.sleep(0.1)
+
+    auth_url = (
+        f"https://{SALESFORCE_DOMAIN}/services/oauth2/authorize?"
+        f"response_type=code&client_id={SALESFORCE_CLIENT_ID}&"
+        f"redirect_uri={SALESFORCE_REDIRECT_URI}&"
+        f"state={state}&nonce={nonce}&"
+        f"code_challenge={code_challenge}&code_challenge_method=S256"
+    )
+    webbrowser.open(auth_url)
+
+    timeout = 60
+    start_time = time.time()
+    while time.time() - start_time < timeout:
+        with server_lock:
+            if auth_code and received_state:
+                break
+        time.sleep(0.1)
+    else:
+        raise TimeoutError("Timeout waiting for callback.")
+    server_thread.join()
+
+    with server_lock:
+        if received_state != state:
+            raise ValueError("State mismatch.")
+        if not auth_code:
+            raise ValueError("No authorization code received")
+
+        token_url = f"https://{SALESFORCE_DOMAIN}/services/oauth2/token"
+        token_data = {
+            "grant_type": "authorization_code",
+            "client_id": SALESFORCE_CLIENT_ID,
+            "client_secret": SALESFORCE_CLIENT_SECRET,
+            "redirect_uri": SALESFORCE_REDIRECT_URI,
+            "code": auth_code,
+            "code_verifier": code_verifier,
+        }
+
```

Comment:
Is there a library from Salesforce to do this instead of using the raw REST API?